plan/statistics/statistics.go

```diff
@@ -57,13 +68,101 @@ func (c *Column) String() string {
 	return strings.Join(strs, "\n")
 }
 
+// EqualRowCount estimates the row count where the column equals to value.
+func (c *Column) EqualRowCount(value types.Datum) (int64, error) {
+	if len(c.Numbers) == 0 {
+		return pseudoRowCount / pseudoEqualRate, nil
+	}
+	index, match, err := c.search(value)
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+	if index == len(c.Numbers) {
+		return c.Numbers[index-1] + 1, nil
+	}
+	if match {
+		return c.Repeats[index] + 1, nil
+	}
+	totalCount := c.Numbers[len(c.Numbers)-1] + 1
+	return totalCount / c.NDV, nil
+}
+
+// LessRowCount estimates the row count where the column less than value.
+func (c *Column) LessRowCount(value types.Datum) (int64, error) {
+	if len(c.Numbers) == 0 {
+		return pseudoRowCount / pseudoLessRate, nil
+	}
+	index, match, err := c.search(value)
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+	if index == len(c.Numbers) {
+		return c.totalRowCount(), nil
+	}
+	number := c.Numbers[index]
+	prevNumber := int64(0)
+	if index > 0 {
+		prevNumber = c.Numbers[index-1]
+	}
+	lessThanBucketValueCount := number - c.Repeats[index]
+	if match {
+		return lessThanBucketValueCount, nil
+	}
+	return (prevNumber + lessThanBucketValueCount) / 2, nil
+}
+
+// BetweenRowCount estimate the row count where column greater or equal to a and less than b.
+func (c *Column) BetweenRowCount(a, b types.Datum) (int64, error) {
+	if len(c.Numbers) == 0 {
+		return pseudoRowCount / pseudoBetweenRate, nil
+	}
+	lessCountA, err := c.LessRowCount(a)
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+	lessCountB, err := c.LessRowCount(b)
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+	if lessCountA >= lessCountB {
+		return c.inBucketBetweenCount(), nil
+	}
+	return lessCountB - lessCountA, nil
+}
+
+func (c *Column) totalRowCount() int64 {
+	return c.Numbers[len(c.Numbers)-1] + 1
+}
+
+func (c *Column) bucketRowCount() int64 {
+	return c.totalRowCount() / int64(len(c.Numbers))
+}
+
+func (c *Column) inBucketBetweenCount() int64 {
+	return c.bucketRowCount()/3 + 1
+}
+
+func (c *Column) search(target types.Datum) (index int, match bool, err error) {
+	index = sort.Search(len(c.Values), func(i int) bool {
+		cmp, err1 := c.Values[i].CompareDatum(target)
+		if err1 != nil {
+			err = errors.Trace(err1)
+			return false
+		}
+		if cmp == 0 {
+			match = true
+		}
+		return cmp >= 0
+	})
+	return
+}
+
 // Table represents statistics for a table.
 type Table struct {
-	info        *model.TableInfo
-	TS          int64 // build timestamp.
-	Columns     []*Column
-	Count       int64 // Total row count in a table.
-	BucketCount int64 // Number of histogram bucket.
+	info    *model.TableInfo
+	TS      int64 // build timestamp.
```

Comment:
They may be used in other packages, make it public so they can be used anyway user want to.